include/toml++/impl/path.h

```diff
@@ -0,0 +1,411 @@
+	/// \ecpp
+	///
+	/// \out
+	/// second cat: lion
+	/// cats: ['tiger', 'lion', 'puma'] 
+	/// \eout
+	class TOML_EXPORTED_CLASS path
+	{
+	  private:
+		/// \cond
+
+		bool parse_error_ = false;
+
+		std::vector<path_component> components_;
+
+		std::vector<path_component> parse_(std::string_view, bool& parse_success);
+
+		/// \endcond
+
+	  public:
+		/// \brief	Default constructor.
+		TOML_NODISCARD_CTOR
+		TOML_EXPORTED_MEMBER_FUNCTION
+		path() noexcept = default;
+
+		/// \brief Construct from string
+		TOML_NODISCARD_CTOR
+		TOML_EXPORTED_MEMBER_FUNCTION
+		explicit path(std::string_view);
+
+		/// \brief	Default destructor.
+		TOML_EXPORTED_MEMBER_FUNCTION
+		~path() noexcept = default;
+
+		/// \brief	Copy constructor.
+		TOML_NODISCARD_CTOR
+		TOML_EXPORTED_MEMBER_FUNCTION
+		inline path(const path& other) { *this = other; }
+
+		/// \brief	Move constructor.
+		TOML_NODISCARD_CTOR
+		TOML_EXPORTED_MEMBER_FUNCTION
+		path(path&& other) noexcept { *this = std::move(other); }
+
+		/// \brief	Copy-assignment operator.
+		TOML_EXPORTED_MEMBER_FUNCTION
+		path& operator=(const path&);
+
+		/// \brief	Move-assignment operator.
+		TOML_EXPORTED_MEMBER_FUNCTION
+		path& operator=(path&&) noexcept;
+
+		/// \brief	Append operator.
+		TOML_EXPORTED_MEMBER_FUNCTION
+		path& operator/=(path&&) noexcept;
+
+		/// \brief	Append operator.
+		TOML_EXPORTED_MEMBER_FUNCTION
+		path& operator/=(std::string_view);
+
+		/// \brief	Append operator.
+		TOML_EXPORTED_MEMBER_FUNCTION
+		path& operator+=(path&&) noexcept;
+
+		/// \brief	Append operator.
+		TOML_EXPORTED_MEMBER_FUNCTION
+		path& operator+=(std::string_view);
+
+		/// \brief Evaluate whether path parsing succeeded
+		inline operator bool() const noexcept
+		{
+			return !parse_error_;
+		};
+
+		/// \brief Implicitly cast to a std::string
+		inline operator std::string() const { return string(); }
+
+		/// \brief Evaluate whether two paths are the same
+		TOML_EXPORTED_MEMBER_FUNCTION
+		bool operator==(const path& compare) const noexcept;
+
+		/// \brief Evaluate whether two paths are the same
+		TOML_EXPORTED_MEMBER_FUNCTION
+		bool operator==(std::string_view compare) const noexcept;
+
+		/// \brief Evaluate whether two paths are the same
+		TOML_EXPORTED_MEMBER_FUNCTION
+		bool operator==(const char* compare) const noexcept;
+
+		/// \brief Evaluate whether two paths are different
+		TOML_EXPORTED_MEMBER_FUNCTION
+		bool operator!=(const path& compare) const noexcept;
+
+		/// \brief Evaluate whether two paths are different
+		TOML_EXPORTED_MEMBER_FUNCTION
+		bool operator!=(std::string_view compare) const noexcept;
+
+		/// \brief Evaluate whether two paths are different
+		TOML_EXPORTED_MEMBER_FUNCTION
+		bool operator!=(const char* compare) const noexcept;
+
+		/// \brief Fetch a path component by index
+		inline path_component* operator[](size_t index) noexcept
+		{
+			if (index < components_.size())
+				return &components_[index];
+
+			return nullptr;
+		};
```

Comment:
That's a good point, and I can't remember why I decided to return a pointer here. It might have been to return nullptr if doesn't exist? Could just return an empty path_component with `.type = path_component_type::invalid` though.

Could certainly do some OOP stuff to maintain value/type relationship, though what I did with the variant is that at_path just checks that the contained type is sensible before returning, and if not returns an empty node_view.